[Android] Native UI Component does not render anything.

# Details

**Current**
The PSPDFKitView Component does not render anything on Android.

**Expected**
The PdfView is rendered in the view hierarchy as it is on ios.

**Additional Information**

tested with react-native-pspdfkit version: 1.24.2 and 1.23.7

We tried to use paths formatted like this: `/data/user/0/com.appname/files/tmp/dokument/7_0` and `file:///data/user/0/com.appname/files/tmp/dokument/7_0`

We are using `ReactFragmentActivity` as documented in the Readme.
```
import com.facebook.react.ReactFragmentActivity;

public class MainActivity extends ReactFragmentActivity {
    /**
     * Returns the name of the main component registered from JavaScript.
     * This is used to schedule rendering of the component.
     */
    @Override
    protected String getMainComponentName() {
        return "my-appname";
    }
}
```

# Acceptance Criteria

- [ ] PSPDFKitView renders the pdf supplied through the `document prop`
- [ ] If the value of the `document` prop changes, the new pdf will be rendered.

@irgendeinich 

We updated our demo application to support android. The steps to reproduce the issue are the following:

1. clone https://github.com/leoek/react-native-pspdfkit-demo
2. checkout reproduce-isssue-237_native-ui-comp-not-rendering
3. `yarn`
4. `yarn android`
5. Press "Go To PdfScreen"
--> There should be this message (from within the render function) in the log now 
```
["pdf"] "render pdfview" "/data/user/0/com.pspdfkitdemo/files/sitzungsunterlage_0_0"
```
--> **Current**: no pdf is rendered
--> **Expected**: PdfViewer with loremipsum pdf is rendered

6. Click on "Back" in the top left
7. Click on "Go To PdfScreen"

--> **Current**: The App will crash with the following stacktrace:
```
➜ adb logcat -s System.err:* om-pspdfkitdem:*
--------- beginning of main
--------- beginning of system
07-02 15:04:32.934  6107  6107 W System.err: io.reactivex.exceptions.OnErrorNotImplementedException: Error while loading PdfDocument
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.internal.observers.ConsumerSingleObserver.onError(ConsumerSingleObserver.java:47)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:79)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:119)
07-02 15:04:32.934  6107  6107 W System.err:    at android.os.Handler.handleCallback(Handler.java:873)
07-02 15:04:32.934  6107  6107 W System.err:    at android.os.Handler.dispatchMessage(Handler.java:99)
07-02 15:04:32.934  6107  6107 W System.err:    at android.os.Looper.loop(Looper.java:193)
07-02 15:04:32.934  6107  6107 W System.err:    at android.app.ActivityThread.main(ActivityThread.java:6669)
07-02 15:04:32.934  6107  6107 W System.err:    at java.lang.reflect.Method.invoke(Native Method)
07-02 15:04:32.934  6107  6107 W System.err:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
07-02 15:04:32.934  6107  6107 W System.err:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
07-02 15:04:32.934  6107  6107 W System.err: Caused by: java.io.IOException: Error while loading PdfDocument
07-02 15:04:32.934  6107  6107 W System.err:    at com.pspdfkit.framework.fz.p(SourceFile:432)
07-02 15:04:32.934  6107  6107 W System.err:    at com.pspdfkit.framework.fz.<init>(SourceFile:333)
07-02 15:04:32.934  6107  6107 W System.err:    at com.pspdfkit.framework.fz.a(SourceFile:124)
07-02 15:04:32.934  6107  6107 W System.err:    at com.pspdfkit.framework.da$1.call(SourceFile:1059)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.Single.subscribe(Single.java:3394)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.internal.operators.single.SingleMap.subscribeActual(SingleMap.java:34)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.Single.subscribe(Single.java:3394)
07-02 15:04:32.934  6107  6107 W System.err:    at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
07-02 15:04:32.935  6107  6107 W System.err:    at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:579)
07-02 15:04:32.935  6107  6107 W System.err:    at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
07-02 15:04:32.935  6107  6107 W System.err:    at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
07-02 15:04:32.935  6107  6107 W System.err:    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
07-02 15:04:32.935  6107  6107 W System.err:    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
07-02 15:04:32.935  6107  6107 W System.err:    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
07-02 15:04:32.935  6107  6107 W System.err:    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
07-02 15:04:32.935  6107  6107 W System.err:    at java.lang.Thread.run(Thread.java:764)
07-02 15:04:32.935  6107  6107 W System.err: Caused by: java.lang.RuntimeException: File not found or could not be opened. (errno:2, file://data/user/0/com.pspdfkitdemo/files/sitzungsunterlage_0_0)
07-02 15:04:32.935  6107  6107 W System.err:    at com.pspdfkit.framework.jni.NativeDocument.open(Native Method)
07-02 15:04:32.935  6107  6107 W System.err:    at com.pspdfkit.framework.fz.p(SourceFile:419)
07-02 15:04:32.935  6107  6107 W System.err:    ... 16 more
```
--> **Expected**: PdfViewer with loremipsum pdf is rendered

--- a/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
+++ b/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
@@ -16,6 +16,7 @@ import com.pspdfkit.preferences.PSPDFKitPreferences;
 import com.pspdfkit.react.events.PdfViewAnnotationChangedEvent;
 import com.pspdfkit.react.events.PdfViewAnnotationTappedEvent;
 import com.pspdfkit.react.events.PdfViewDataReturnedEvent;
+import com.pspdfkit.react.events.PdfViewDocumentLoadFailedEvent;
 import com.pspdfkit.react.events.PdfViewDocumentSaveFailedEvent;
 import com.pspdfkit.react.events.PdfViewDocumentSavedEvent;
 import com.pspdfkit.react.events.PdfViewStateChangedEvent;
@@ -148,11 +149,13 @@ public class ReactPdfViewManager extends ViewGroupManager<PdfView> {
     @Override
     public Map getExportedCustomDirectEventTypeConstants() {
         return MapBuilder.of(PdfViewStateChangedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onStateChanged"),
-                PdfViewDocumentSavedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentSaved"),
-                PdfViewAnnotationTappedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onAnnotationTapped"),
-                PdfViewAnnotationChangedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onAnnotationsChanged"),
-                PdfViewDataReturnedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDataReturned"),
-                PdfViewDocumentSaveFailedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentSaveFailed"));
+            PdfViewDocumentSavedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentSaved"),
+            PdfViewAnnotationTappedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onAnnotationTapped"),
+            PdfViewAnnotationChangedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onAnnotationsChanged"),
+            PdfViewDataReturnedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDataReturned"),
+            PdfViewDocumentSaveFailedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentSaveFailed"),
+            PdfViewDocumentLoadFailedEvent.EVENT_NAME, MapBuilder.of("registrationName", "onDocumentLoadFailed")
+        );
     }
 
     @Override

--- a/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
+++ b/android/src/main/java/com/pspdfkit/react/ReactPdfViewManager.java
@@ -76,7 +76,7 @@ public class ReactPdfViewManager extends ViewGroupManager<PdfView> {
 
     @Override
     public void onDropViewInstance(PdfView view) {
-        view.removeFragment();
+        view.removeFragment(true);
     }
 
     @Nullable

--- a/android/src/main/java/com/pspdfkit/react/events/PdfViewDocumentLoadFailedEvent.java
+++ b/android/src/main/java/com/pspdfkit/react/events/PdfViewDocumentLoadFailedEvent.java
@@ -1,0 +1,36 @@
+package com.pspdfkit.react.events;
+
+import android.support.annotation.IdRes;
+import android.support.annotation.NonNull;
+
+import com.facebook.react.bridge.Arguments;
+import com.facebook.react.bridge.WritableMap;
+import com.facebook.react.uimanager.events.Event;
+import com.facebook.react.uimanager.events.RCTEventEmitter;
+
+/**
+ * Event sent by the {@link com.pspdfkit.views.PdfView} when the document load failed.
+ */
+public class PdfViewDocumentLoadFailedEvent extends Event<PdfViewDocumentLoadFailedEvent> {
+
+    public static final String EVENT_NAME = "pdfViewDocumentLoadFailed";
+
+    private final String error;
+
+    public PdfViewDocumentLoadFailedEvent(@IdRes int viewId, @NonNull String error) {
+        super(viewId);
+        this.error = error;
+    }
+
+    @Override
+    public String getEventName() {
+        return EVENT_NAME;
+    }
+
+    @Override
+    public void dispatch(RCTEventEmitter rctEventEmitter) {
+        WritableMap eventData = Arguments.createMap();
+        eventData.putString("error", error);
+        rctEventEmitter.receiveEvent(getViewTag(), getEventName(), eventData);
+    }
+}

--- a/android/src/main/java/com/pspdfkit/views/PdfView.java
+++ b/android/src/main/java/com/pspdfkit/views/PdfView.java
@@ -153,6 +153,9 @@ public class PdfView extends FrameLayout {
                 }
             }
         });
+
+        // Set a default configuration.
+        configuration = new PdfActivityConfiguration.Builder(getContext()).build();
     }
 
     public void inject(FragmentManager fragmentManager, EventDispatcher eventDispatcher) {
@@ -172,7 +175,12 @@ public class PdfView extends FrameLayout {
         setupFragment();
     }
 
-    public void setDocument(String document) {
+    public void setDocument(@Nullable String document) {
+        if (document == null) {
+            removeFragment(false);
+            return;
+        }
+
         if (Uri.parse(document).getScheme() == null) {
             // If there is no scheme it might be a raw path.
             try {
@@ -317,16 +325,19 @@ public class PdfView extends FrameLayout {
         });
     }
 
-    public void removeFragment() {
+    public void removeFragment(boolean makeInactive) {
         PdfFragment pdfFragment = (PdfFragment) fragmentManager.findFragmentByTag(fragmentTag);
         if (pdfFragment != null) {
             fragmentManager.beginTransaction()
                     .remove(pdfFragment)
                     .commitAllowingStateLoss();
         }
-        isActive = false;
+        if (makeInactive) {
+            isActive = false;
+        }
 
         fragment = null;
+        document = null;
         fragmentGetter.onComplete();
         fragmentGetter = BehaviorSubject.create();
         pendingFragmentActions.dispose();
@@ -335,6 +346,7 @@ public class PdfView extends FrameLayout {
             textSelectionPopupToolbar.dismiss();
             textSelectionPopupToolbar = null;
         }
+        pdfThumbnailBar.setVisibility(View.GONE);
     }
 
     void manuallyLayoutChildren() {

--- a/android/src/main/java/com/pspdfkit/views/PdfView.java
+++ b/android/src/main/java/com/pspdfkit/views/PdfView.java
@@ -31,6 +31,7 @@ import com.pspdfkit.listeners.OnPreparePopupToolbarListener;
 import com.pspdfkit.listeners.SimpleDocumentListener;
 import com.pspdfkit.react.R;
 import com.pspdfkit.react.events.PdfViewDataReturnedEvent;
+import com.pspdfkit.react.events.PdfViewDocumentLoadFailedEvent;
 import com.pspdfkit.react.events.PdfViewDocumentSaveFailedEvent;
 import com.pspdfkit.react.events.PdfViewDocumentSavedEvent;
 import com.pspdfkit.react.events.PdfViewStateChangedEvent;
@@ -49,6 +50,7 @@ import org.json.JSONException;
 import org.json.JSONObject;
 
 import java.io.ByteArrayOutputStream;
+import java.io.File;
 import java.util.ArrayList;
 import java.util.EnumSet;
 import java.util.List;
@@ -172,7 +174,13 @@ public class PdfView extends FrameLayout {
 
     public void setDocument(String document) {
         if (Uri.parse(document).getScheme() == null) {
-            document = FILE_SCHEME + document;
+            // If there is no scheme it might be a raw path.
+            try {
+                File file = new File(document);
+                document = Uri.fromFile(file).toString();
+            } catch (Exception e) {
+                document = FILE_SCHEME + document;
+            }
         }
         if (documentOpeningDisposable != null) {
             documentOpeningDisposable.dispose();
@@ -187,6 +195,14 @@ public class PdfView extends FrameLayout {
                     public void accept(PdfDocument pdfDocument) throws Exception {
                         PdfView.this.document = pdfDocument;
                         setupFragment();
+                    }
+
+                }, new Consumer<Throwable>() {
+                    @Override
+                    public void accept(Throwable throwable) throws Exception {
+                        PdfView.this.document = null;
+                        setupFragment();
+                        eventDispatcher.dispatchEvent(new PdfViewDocumentLoadFailedEvent(getId(), throwable.getMessage()));
                     }
                 });
     }

--- a/index.js
+++ b/index.js
@@ -35,6 +35,7 @@ class PSPDFKitView extends React.Component {
           onStateChanged={this._onStateChanged}
           onDocumentSaved={this._onDocumentSaved}
           onDocumentSaveFailed={this._onDocumentSaveFailed}
+          onDocumentLoadFailed={this._onDocumentLoadFailed}
           onAnnotationTapped={this._onAnnotationTapped}
           onAnnotationsChanged={this._onAnnotationsChanged}
           onDataReturned={this._onDataReturned}
@@ -60,6 +61,12 @@ class PSPDFKitView extends React.Component {
   _onDocumentSaveFailed = event => {
     if (this.props.onDocumentSaveFailed) {
       this.props.onDocumentSaveFailed(event.nativeEvent);
+    }
+  };
+
+  _onDocumentLoadFailed = event => {
+    if (this.props.onDocumentLoadFailed) {
+      this.props.onDocumentLoadFailed(event.nativeEvent);
     }
   };
 

--- a/ios/RCTPSPDFKit/RCTPSPDFKitView.h
+++ b/ios/RCTPSPDFKit/RCTPSPDFKitView.h
@@ -26,6 +26,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, copy) RCTBubblingEventBlock onCloseButtonPressed;
 @property (nonatomic, copy) RCTBubblingEventBlock onDocumentSaved;
 @property (nonatomic, copy) RCTBubblingEventBlock onDocumentSaveFailed;
+@property (nonatomic, copy) RCTBubblingEventBlock onDocumentLoadFailed;
 @property (nonatomic, copy) RCTBubblingEventBlock onAnnotationTapped;
 @property (nonatomic, copy) RCTBubblingEventBlock onAnnotationsChanged;
 @property (nonatomic, copy) RCTBubblingEventBlock onStateChanged;

--- a/ios/RCTPSPDFKit/RCTPSPDFKitView.m
+++ b/ios/RCTPSPDFKit/RCTPSPDFKitView.m
@@ -13,7 +13,7 @@
 #import "RCTConvert+PSPDFViewMode.h"
 #import "RCTConvert+UIBarButtonItem.h"
 
-#define VALIDATE_DOCUMENT(document, ...) { if (document.isValid) { NSLog(@"Document is invalid."); if (self.onDocumentLoadFailed) { self.onDocumentLoadFailed(@{@"error": @"Document is invalid."}); } return __VA_ARGS__; }}
+#define VALIDATE_DOCUMENT(document, ...) { if (!document.isValid) { NSLog(@"Document is invalid."); if (self.onDocumentLoadFailed) { self.onDocumentLoadFailed(@{@"error": @"Document is invalid."}); } return __VA_ARGS__; }}
 
 @interface RCTPSPDFKitView ()<PSPDFDocumentDelegate, PSPDFViewControllerDelegate, PSPDFFlexibleToolbarContainerDelegate>
 

--- a/ios/RCTPSPDFKit/RCTPSPDFKitView.m
+++ b/ios/RCTPSPDFKit/RCTPSPDFKitView.m
@@ -13,7 +13,7 @@
 #import "RCTConvert+PSPDFViewMode.h"
 #import "RCTConvert+UIBarButtonItem.h"
 
-#define VALIDATE_DOCUMENT(document, ...) { if (!document.isValid) { NSLog(@"Document is invalid."); return __VA_ARGS__; }}
+#define VALIDATE_DOCUMENT(document, ...) { if (document.isValid) { NSLog(@"Document is invalid."); if (self.onDocumentLoadFailed) { self.onDocumentLoadFailed(@{@"error": @"Document is invalid."}); } return __VA_ARGS__; }}
 
 @interface RCTPSPDFKitView ()<PSPDFDocumentDelegate, PSPDFViewControllerDelegate, PSPDFFlexibleToolbarContainerDelegate>
 
@@ -158,6 +158,10 @@
 
 - (void)pdfViewController:(PSPDFViewController *)pdfController willBeginDisplayingPageView:(PSPDFPageView *)pageView forPageAtIndex:(NSInteger)pageIndex {
   [self onStateChangedForPDFViewController:pdfController pageView:pageView pageAtIndex:pageIndex];
+}
+
+- (void)pdfViewController:(PSPDFViewController *)pdfController didChangeDocument:(nullable PSPDFDocument *)document {
+    VALIDATE_DOCUMENT(document)
 }
 
 #pragma mark - PSPDFFlexibleToolbarContainerDelegate

--- a/ios/RCTPSPDFKit/RCTPSPDFKitViewManager.m
+++ b/ios/RCTPSPDFKit/RCTPSPDFKitViewManager.m
@@ -93,6 +93,8 @@ RCT_EXPORT_VIEW_PROPERTY(onDocumentSaved, RCTBubblingEventBlock)
 
 RCT_EXPORT_VIEW_PROPERTY(onDocumentSaveFailed, RCTBubblingEventBlock)
 
+RCT_EXPORT_VIEW_PROPERTY(onDocumentLoadFailed, RCTBubblingEventBlock)
+
 RCT_EXPORT_VIEW_PROPERTY(onAnnotationTapped, RCTBubblingEventBlock)
 
 RCT_EXPORT_VIEW_PROPERTY(onAnnotationsChanged, RCTBubblingEventBlock)

